File: javascript/src/AddTwoNumbers.js
// Problem ID: 2
// URL: https://leetcode.com/problems/add-two-numbers/

const { ListNode } = require('../helper/linkedList');

const solution = (l1, l2) => {
    let carry = 0;
    let result;
    let currentNode;
    while (l1 || l2 || carry) {
        let sum = (l1 ? l1.val : 0) + (l2 ? l2.val : 0) + carry
        if (sum >= 10) {
            sum %= 10;
            carry = 1;
        } else {
            carry = 0;
        }
        let node = new ListNode(sum);
        if (!result) {
            result = currentNode = node;
        } else {
            currentNode.next = node;
            currentNode = node;
        }
        if (l1) l1 = l1.next;
        if (l2) l2 = l2.next;
    }
    
    return result;
}

const solution2 = (l1, l2, c = 0) => {
    if (!l1 && !l2) return c ? new ListNode(c) : null;
    if (l1 && !l2) return new ListNode((l1.val + c) % 10, solution2(l1.next, null, Math.floor((l1.val + c) / 10)))
    if (!l1 && l2) return new ListNode((l2.val + c) % 10, solution2(null, l2.next, Math.floor((l2.val + c) / 10)))
    return new ListNode((l1.val + l2.val + c) % 10, solution2(l1.next, l2.next, Math.floor((l1.val + l2.val + c) / 10)))
}

module.exports = {
    solution,
    solution2,
};

File: scala/src/main/scala/AddTwoNumbers.scala
package leetcode

// https://leetcode.com/problems/add-two-numbers/discuss/1056/Scala-solution
object AddTwoNumbers {
    def addTwoNumbers(l1: List[Int], l2: List[Int]): List[Int] = {        
        def f(lists: (List[Int], List[Int]), carry: Int) = lists match {
            case (Nil, Nil) => if (carry == 0) Nil else List(carry)
            case (x::xtail, Nil) => g(x, 0, carry, (xtail, Nil))
            case (Nil, y::ytail) => g(0, y, carry, (Nil, ytail))
            case (x::xtail, y::ytail) => g(x, y, carry, (xtail, ytail))
        }

        def g(x: Int, y: Int, carry: Int, tails: (List[Int], List[Int])): List[Int] = {
            val sum = x + y + carry
            sum % 10 :: f(tails, sum / 10)
        }
        f((l1, l2), 0)
    }
}

// https://leetcode.com/problems/add-two-numbers/discuss/1074/Yet-another-scala-solution
object AddTwoNumbers2 {
    def addTwoNumbers(l1: List[Int], l2: List[Int]): List[Int] = addTwoNumbers(l1, l2, 0)
    
    def addTwoNumbers(l1: List[Int], l2: List[Int], c: Int): List[Int] = (l1, l2, c) match {
        case (Nil, Nil, 0) => Nil
        case (Nil, Nil, _) => List(c)
        case (x::xt, Nil, _) => (x + c) % 10 :: addTwoNumbers(xt, Nil, (x + c) / 10)
        case (Nil, y::yt, _) => (y + c) % 10 :: addTwoNumbers(Nil, yt, (y + c) / 10)
        case (x::xt, y::yt, _) => (x + y + c) % 10 :: addTwoNumbers(xt, yt, (x + y + c) / 10)
    }
}

object AddTwoNumbers3 {
    import ListNode.{listNodeToList, listToListNode}

    def addTwoNumbers(l1: ListNode, l2: ListNode): List[Int] = addTwoNumbers(l1, l2, 0)
    def addTwoNumbers(l1: List[Int], l2: List[Int], c: Int): List[Int] = (l1, l2, c) match {
        case (Nil, Nil, 0) => Nil
        case (Nil, Nil, _) => List(c)
        case (h1::t1, Nil, _) => (h1 + c) % 10 :: addTwoNumbers(t1, Nil, (h1 + c) / 10)
        case (Nil, h2::t2, _) => (h2 + c) % 10 :: addTwoNumbers(Nil, t2, (h2 + c) % 10)
        case (h1::t1, h2::t2, _) => (h1 + h2 + c) % 10 :: addTwoNumbers(t1, t2, (h1 + h2 + c) / 10)
    }
}
